index.js

```diff
@@ -3232,6 +3232,57 @@
   S.range =
   def('range', {}, [$.Integer, $.Integer, $.Array($.Integer)], range);
 
+  //# groupBy :: (a -> a -> Boolean) -> Array a -> Array (Array a)
+  //.
+  //. Splits its array argument into an array of arrays of equal,
+  //. adjacent elements. The equality of the elements is tested with
+  //. a curried function provided by the user in the first argument.
+  //.
+  //. The behaviour of groupBy can be surprising for functions that
+  //. aren't reflexive, transitive and symmetric
+  //. (see [equivalence][] relation).
+
+  //. See also [`groupBy_`](#groupBy_).
+  //.
+  //. ```javascript
+  //. > S.groupBy(S.equals, [1, 1, 2, 3, 3, 4])
+  //. [[1, 1], [2], [3, 3], [4]]
+  //.
+  //. > S.groupBy(S.equals, [1, 1, 2, 1, 1])
+  //. [[1, 1], [2], [1, 1]]
+  //.
+  //. > S.groupBy(a => b => a * b % 3 === 0, [1, 2, 3, 4, 5, 6, 7, 8, 9])
+  //. [[1], [2, 3], [4], [5, 6], [7], [8, 9]]
+  //. ```
+  function groupBy(f, xs) {
+    return groupBy_(uncurry2(f), xs);
+  }
+  S.groupBy =
+  def('groupBy',
+      {},
+      [Fn(a, Fn(a, $.Boolean)), $.Array(a), $.Array($.Array(a))],
+      groupBy);
+
+  //# groupBy_ :: ((a, a) -> Boolean) -> Array a -> Array (Array a)
+  //.
+  //. Variant of [`groupBy`](#groupBy) which takes an uncurried function.
+  function groupBy_(f, xs) {
+    if (xs.length === 0) return [];
+    var x0 = xs[0];         // :: a
+    var active = [x0];      // :: Array a
+    var result = [active];  // :: Array (Array a)
+    for (var idx = 1; idx < xs.length; idx += 1) {
+      var x = xs[idx];
+      if (f(x0, x)) active.push(x); else result.push(active = [x0 = x]);
```

Comment:
I prefer `if (...) ...; else ...;` for exactly the reason you suggested, Stefano.